bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt

```diff
@@ -0,0 +1,124 @@
+import com.r3.businessnetworks.membership.flows.member.MembershipsListResponse
+import com.r3.businessnetworks.membership.flows.member.RequestMembershipFlow
+import com.r3.businessnetworks.membership.flows.member.service.MembershipsCacheHolder
+import com.r3.businessnetworks.membership.states.MembershipContract
+import com.r3.businessnetworks.membership.states.MembershipState
+
+
+import com.r3.businessnetworks.memberships.demo.contracts.SampleContract
+import com.r3.businessnetworks.memberships.demo.contracts.SampleState
+import net.corda.core.contracts.ReferencedStateAndRef
+import net.corda.core.contracts.StateAndRef
+import net.corda.core.contracts.requireThat
+import net.corda.core.flows.*
+import net.corda.core.identity.Party
+import net.corda.core.serialization.CordaSerializable
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+import net.corda.core.utilities.unwrap
+
+@CordaSerializable
+class RequestMembershipForTransaction (val bno : Party)
+
+class IssueAssetFlow (val bno : Party) : FlowLogic<SignedTransaction>() {
+    @Suspendable
+    override fun call() : SignedTransaction {
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // query the vault with bno
+        val memberships= subFlow(GetMembershipsFlow(bno))
+        val ourMembership = memberships[ourIdentity]!!
+
+        val builder = TransactionBuilder(notary)
+                .addOutputState(SampleState(ourIdentity), SampleContract.CONTRACT_NAME)
+                .addCommand(SampleContract.Commands.Issue(), ourIdentity.owningKey)
+                .addReferenceState(ReferencedStateAndRef(ourMembership))
+
+        builder.verify(serviceHub)
+        val stx = serviceHub.signInitialTransaction(builder)
+        return subFlow(FinalityFlow(stx, listOf()))
+    }
+}
+
+@InitiatingFlow
+class TransferAssetFlow(val sampleState : StateAndRef<SampleState>,
+                              val bno : Party,
+                              val partyToTransferTo : Party) : FlowLogic<SignedTransaction>() {
+    @Suspendable
+    override fun call() : SignedTransaction {
+        if (ourIdentity != sampleState.state.data.owner)
+            throw FlowException("Only owner of the state can transfer it")
+        if (ourIdentity == partyToTransferTo)
+            throw FlowException("sender and recipient should be different parties")
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // get the membership from the node's membership cache
+        val membershipsCacheHolderService = serviceHub.cordaService(MembershipsCacheHolder::class.java)
+        val ourMembership = membershipsCacheHolderService.cache.getMembership(bno, ourIdentity)
```

Comment:
We should also verify that `partyToTransferTo` has membership as well

Reply:
@ivanr3 
Right now, in the TransferAssetFlow this is checked in AssetContract.Commands.Transfer() when the verify() runs.
The membership for the counterparty is obtained for the counterparty as follows (without the ReceiveStateAndRefFlow the chain resolution fails - as it should):
        // create a session
        val counterPartySession = initiateFlow(partyToTransferTo)
        // requesting other party to send us their membership state
        counterPartySession.send(RequestMembershipForTransaction(bno))
        val counterPartyMembershipState =
                subFlow(ReceiveStateAndRefFlow<MembershipState<Any>>(counterPartySession)).single()

Question: On top of the check above, should the membership of the counterparty be checked in the TransferAssetFlow as follows : subFlow(GetMembershipsFlow(bno))[partyToTransferTo]
Is that what you have in mind?

Am I missing sth?